cs3api4lab/cs3_file_api.py

```diff
@@ -0,0 +1,237 @@
+
+        time_end = time.time()
+        self.log.info('msg="Invoked stat" fileid="%s" elapsedTimems="%.1f"' % (fileid, (time_end - time_start) * 1000))
+
+        if stat_info.status.code == cs3code.CODE_OK:
+            self.log.debug('msg="Stat result" data="%s"' % stat_info)
+            return {
+                'inode': stat_info.info.id.storage_id + ':' + stat_info.info.id.opaque_id,
+                'filepath': stat_info.info.path,
+                'userid': stat_info.info.owner.opaque_id,
+                'size': stat_info.info.size,
+                'mtime': stat_info.info.mtime.seconds
+            }
+
+        self.log.info('msg="Failed stat" fileid="%s" reason="%s"' % (fileid, stat_info.status.message))
+        raise IOError(stat_info.status.message)
+
+    def read_file(self, endpoint, filepath, userid):
+        """
+        Read a file using the given userid as access token.
+        """
+
+        time_start = time.time()
+
+        #
+        # Prepare endpoint
+        #
+        reference = self._cs3_reference(endpoint, filepath)
+
+        req = cs3sp.InitiateFileDownloadRequest(ref=reference)
+
+        init_file_download = self.cs3_stub.InitiateFileDownload(request=req, metadata=[('x-access-token', self._authenticate(userid))])
+
+        if init_file_download.status.code == cs3code.CODE_NOT_FOUND:
+            self.log.info('msg="File not found on read" filepath="%s"' % filepath)
+            yield IOError('No such file or directory')
+
+        elif init_file_download.status.code != cs3code.CODE_OK:
+            self.log.debug('msg="Failed to initiateFileDownload on read" filepath="%s" reason="%s"' % \
+                                  (filepath, init_file_download.status.message))
+            yield IOError(init_file_download.status.message)
+
+        self.log.debug('msg="readfile: InitiateFileDownloadRes returned" endpoint="%s"' % init_file_download.download_endpoint)
+
+        #
+        # Download
+        #
+        file_get = None
+        try:
+            file_get = requests.get(url=init_file_download.download_endpoint, headers={'x-access-token': self._authenticate(userid)})
+        except requests.exceptions.RequestException as e:
+            self.log.error('msg="Exception when downloading file from Reva" reason="%s"' % e)
+            yield IOError(e)
+
+        time_end = time.time()
+        data = file_get.content
+
+        if file_get.status_code != http.HTTPStatus.OK:
+            self.log.error('msg="Error downloading file from Reva" code="%d" reason="%s"' % (file_get.status_code, file_get.reason))
+            yield IOError(file_get.reason)
+        else:
+            self.log.info('msg="File open for read" filepath="%s" elapsedTimems="%.1f"' % (filepath, (time_end - time_start) * 1000))
+            for i in range(0, len(data), self.chunksize):
+                yield data[i:i + self.chunksize]
+
+    def write_file(self, endpoint, filepath, userid, content, noversion=0):
```

Comment:
`noversion` could be a boolean

Reply:
"Noversion" attribute is using in WOPI at xrootiface, we can remove it.

Reply:
I saw that you removed it and it's ok, I will merge.
But this is actually something we have interest, since we have a separate component to manage the versioning of the files. But I don't think REVA has - for now - an option to manually create a version, so we cannot do the versioning as the Jupyter expects.